README.md
# GENI Authorization Tool #

This tool will sign credentials that allow a tool to act on behalf of
a user. Aside from direct user interaction, there are two kinds of
entities that interact with this tool: Member Authorities and Tool
frontends.

As a federation-wide utility, this tool will reside at a trusted
location accessible via SSL connection. Any tool frontend can invoke a
JavaScript function which will popup an authorization window for the
user.

## Test Environment ##

To try out GENI Authorization, you will need a GENI certificate and
key bundle in PEM format, and the ability to host the source files on
a webhost. For testing, you need not host them via SSL.

Upload all of the files in the repository to your chosen
destination. Once you have the files hosted, you can visit index.html
by itself to simply load and clear your certificate. Or you can visit
tool.html to try out the example tool and get a signed credential.

## Tool API ##

To use the GENI Authorization Tool in your tool, you will need to
include geni-auth.js into the web frontend of your tool. In your event
function that handles the user's authorization, invoke the
genilib.authorize method.

### genilib.authorize (id, cert, callback) ###

This method pops up an authorization window allowing the user to
create a speaks-for credential for your tool. This should only be
invoked in response to a user-driven event such as a click or the
browser's popup blocker may interfere.

Parameters:

- *id:* An opaque string, usually the URN of the tool
- *cert:* PEM-encoded certificate of the tool. Should be valid for any
  authority your tool needs to establish trust with.
- *callback:* This method will be invoked with a string containing the
  XML GENI credential. There is no timeout and this method may never
  be called if the user closes the authorization window.

Returns: The authorization window

## Alternative Tool API Without Pop Up Window

Alternatively, a simple PHP page index.php accepts HTTP POST from your tool client and renders the authorisation window interface. Upon user completes the authorisation process, it will redirect back to your tool client tool with a proper speaks-for credential attached. This work flow will eliminate the authorisation being launched via a pop up window.

### POST to index.php

Parameters

- *tool_id:* Identifier of the tool
- *backto:* URL that tool would receive speaks-for credential upon successful authorisation.
- *tool_cert:* PEM-encoded certificate of the tool.

### Accept POST in 'backto'

Parameters

- *data_credential:* Speaks-for credential generated upon successful authorisation.

## Member Authority API ##

Member Authorities may optionally present a web interface that lets
the user pass their certificate to the GENI Authorization Tool. If
they do not present a web interface, it will be up to the user to
manually paste in their certificate file.

Once the user's identity has been verified, the member authority
invokes genilib.sendCertificate to pass the certificate to the trusted
signer.

### genilib.sendCertificate (cert) ###

This method sends a user's GENI certificate to the GENI Authorization
Tool which is presumed to live in the browser window that opened the
current one. After sending the message, it will close the current
window or tab. If this function is invoked in some other circumstance,
it simply closes the current window or tab.

Parameters:

- *cert:* A pem-encoded certificate and private key belonging to the
  logged in user.

Returns: Nothing


debian/changelog
xml-signer (1.4) unstable; urgency=low

  * Add a sample config file for CentOS 7

 -- Tom Mitchell <tcmitchell@users.noreply.github.com>  Mon, 14 Nov 2016 16:51:00 +0000

xml-signer (1.3) unstable; urgency=low

  * Add a clear cookies option during MA selection.
  * Add a checkbox to remember authorize decision. Fix formatting.
  * Make the GENI portal stand alone as the prominent account source.
    Emulab.net is demoted to dropdown.

 -- Tom Mitchell <tmitchel@bbn.com>  Mon, 13 Jul 2015 15:44:00 +0000

xml-signer (1.2) unstable; urgency=low

  * Support PKCS#8 private keys

 -- Tom Mitchell <tmitchel@bbn.com>  Fri, 17 Apr 2015 14:06:00 +0000

xml-signer (1.1) unstable; urgency=low

  * Styling changes

 -- Tom Mitchell <tmitchel@bbn.com>  Tue, 31 Mar 2015 18:03:13 +0000

xml-signer (1.0.1) unstable; urgency=low

  * RPM packaging

 -- Tom Mitchell <tmitchel@bbn.com>  Thu, 25 Sep 2014 16:28:00 +0000

xml-signer (1.0) unstable; urgency=low

  * Add new optional authentication protocol
  * Add advanced mode to simplify UI
  * UI Cleanups

 -- Tom Mitchell <tmitchel@bbn.com>  Fri, 19 Sep 2014 12:25:00 +0000

xml-signer (0.9) unstable; urgency=low

  * Preserve newline while posting speaks_for credential
  * Add index.php to install target

 -- Tom Mitchell <tmitchel@bbn.com>  Thu, 07 Aug 2014 15:24:00 +0000

xml-signer (0.8) unstable; urgency=low

  * Allow tools to post id, cert to the signer and redirect back to the
    tool without pop-up

 -- Tom Mitchell <tmitchel@bbn.com>  Tue, 05 Aug 2014 13:07:00 +0000

xml-signer (0.7) unstable; urgency=low

  * Overhaul visual appearance
  * Change default expiration time to 120 days
  * Allow tool to pass in a user key to use by default
  * Add support for unencrypted keys

 -- Tom Mitchell <tmitchel@bbn.com>  Fri, 16 May 2014 12:28:00 +0000

xml-signer (0.6) unstable; urgency=low

  * Switch to forge for encryption and verification
  * Return signer window to caller on genilib.authorize
  * Wording and other UI tweaks

 -- Tom Mitchell <tmitchel@bbn.com>  Mon, 07 Apr 2014 18:31:00 +0000

xml-signer (0.5) unstable; urgency=low

  * Allow an MA to send a credential back to the originating tool
  * Use an external list of member authorities
  * Allow a tool to specify an optional default MA

 -- Tom Mitchell <tmitchel@bbn.com>  Thu, 12 Feb 2014 17:11:00 +0000

xml-signer (0.4) unstable; urgency=low

  * Add KeyValue and the full certificate chain to signed credentials.

 -- Tom Mitchell <tmitchel@bbn.com>  Thu, 22 Jan 2014 17:42:00 +0000

xml-signer (0.3) unstable; urgency=low

  * Use toISOString() for credential expires field

 -- Tom Mitchell <tmitchel@bbn.com>  Thu, 16 Jan 2014 22:05:00 +0000

xml-signer (0.2) unstable; urgency=low

  * Generate ABAC-style credential
  * Initial forge integration

 -- Tom Mitchell <tmitchel@bbn.com>  Mon, 07 Nov 2013 18:25:00 +0000

xml-signer (0.1) unstable; urgency=low

  * Initial release

 -- Tom Mitchell <tmitchel@bbn.com>  Mon, 09 Sep 2013 18:37:51 -0600


xml-signer.js
/*global require: true */

require.config({
  'paths': {
    'underscore': 'lib/underscore-min',
    'jquery': 'require-jquery',
    'bootstrap': 'lib/bootstrap/js/bootstrap.min'
  },
  'shim': {
    'underscore': {
      'exports': '_'
    }
  }
});

require(['jquery', 'underscore', 'error', 'lib/forge/forge',
         'lib/xml-crypto/signed-xml',
         'text!template/credential.txt', 'text!template/no-key.html',
         'text!template/authorize.html', 'bootstrap'],
function ($, _, error, forge, sigExport, xmlText, noKeyText, authorizeText) {
  'use strict';
  var SignedXml = sigExport.SignedXml;

  var singleDay = 1000*60*60*24;

  var saList = [];

  function parseSaList(str)
  {
    saList = [];
    var lines = str.split('\n');
    _.each(lines, function (line) {
      var fields = line.split(' ');
      if (fields.length > 1)
      {
	var name = fields[0].split('+')[1];
	var url;
	if (fields.length === 2)
	{
	  var urlPaths = fields[1].split('/');
	  url = urlPaths[0] + '//' + urlPaths[2].split(':')[0] +
	    '/getsslcertjs.php3';
	}
	else if (fields.length === 3)
	{
	  url = fields[2];
	}
	if (url)
	{
	  saList.push({ name: name,
			url: url });
	}
	else
	{
	  console.log('Invalid URL for ', fields[0]);
	}
      }
    });
  }

  function parseSaListJson(str) {
    $.each(str['identity-providers'], function() {
      saList.push({ name: this.urn.split('+')[1],
        url: this['web-cert-url'],
        icon: this.icon })
    })
  }

  var debugCert = null;
  var toolId = $('script#tool_id').text();
  var speakerCert = null;
  var xmlTemplate = _.template(xmlText);
  var authorizeTemplate = _.template(authorizeText);
  var xml = "";
  var encryptedKey = "";
  var certList = [];
  var certWindow;
  var userId = 'urn:publicid:IDN+jonlab.testbed.emulab.net+user+jld';
  var defaultMA = { url: null, name: null};
  var myCert = $('script#tool_cert').text().replace(/\\n/g, "\n");
  var backTo = $('script#backto').text();

  function initialize()
  {
    var promise = $.get('https://www.emulab.net/protogeni/boot/salist.json');
    promise.then(function (response) {
      parseSaListJson(response);
      var params = getQueryParams(window.location.search);
      if (toolId == '') {
        toolId = params.id;
      }
      if (toolId)
      {
        var data = { ready: true };
        if (window.opener) {
          window.addEventListener('message', messageToolCert, false);
          window.opener.postMessage(data, '*');
	    console.log('posting ready message');
        } else {
          var fake_event = { data: { certificate: myCert, tool: true } }
          messageToolCert(fake_event);
        }
      }
      else
      {
        messageToolCert(null);
      }
    }, function (error) {
      console.log('Error fetching SA List: ', error);
      flagError();
    });
  }

  function messageToolCert(event)
  {
    /* If the tool requested an MA by URL, extract it. */
    if (event && event.data.ma)
    {
      defaultMA.url = event.data.ma.url;
      defaultMA.name = event.data.ma.name;
    }
    if (event && event.data.certificate && event.data.tool)
    {
      debugCert = event.data.certificate;
      var speakerCertList = extractCertificates(event.data.certificate);
      if (speakerCertList.length > 0)
      {
        speakerCert = speakerCertList[0];
      }
//      var cannedCert = $('#preset-bundle');
      var userCert;
      try {
        userCert = JSON.parse(window.localStorage.certificateList)[0];
      } catch (e) {}
      if (! userCert && event.data.userBundle)
      {
	userCert = event.data.userBundle;
      }
/*
      if (! userCert && cannedCert.length > 0)
      {
	userCert = cannedCert[0].textContent;
      }
*/
      if ((toolId && speakerCert)
          || (! toolId && ! speakerCert))
      {
        if (userCert)
        {
          setupClientPem(userCert);
          _.defer(initAuthorize);
        }
        else
        {
          _.defer(initNoKey);
        }
      }
      else
      {
        flagError();
      }
    }
  }

  function flagError()
  {
    $('#main-content').html('<h1>Error</h1><p>There was an error initializing. Close this window and try again.</p>');
  }

  function extractCertificates(pem)
  {
    var inCert = false;
    var userCertificate = "";
    var result = [];
    var lines = pem.split('\n');
    var i = 0;
    for (i = 0; i < lines.length; i += 1)
    {
      if (lines[i] === "-----BEGIN CERTIFICATE-----")
      {
        inCert = true;
        userCertificate = "";
      }
      else if (lines[i] === "-----END CERTIFICATE-----")
      {
        inCert = false;
        result.push(userCertificate);
      }
      else if (inCert)
      {
        userCertificate += lines[i] + '\n';
      }
    }
    return result;
  }

    function wrapCertificates(list)
    {
	var result = "";
	var i = 0;
	for (; i < list.length; i += 1)
	{
	    result += "-----BEGIN CERTIFICATE-----\n";
	    result += list[i];
	    result += "-----END CERTIFICATE-----\n";
	}
	return result;
    }

  function extractKey(pem)
  {
    var pkcs1_begin = "-----BEGIN RSA PRIVATE KEY-----";
    var pkcs1_end   = "-----END RSA PRIVATE KEY-----";
    var pkcs8_begin = "-----BEGIN PRIVATE KEY-----";
    var pkcs8_end   = "-----END PRIVATE KEY-----";
    var inKey = false;
    var key = "";
    var lines = pem.split('\n');
    var i = 0;
    for (i = 0; i < lines.length; i += 1)
    {
      if (lines[i] === pkcs1_begin || lines[i] === pkcs8_begin)
      {
        key = lines[i] + '\n';
        inKey = true;
      }
      else if (lines[i] === pkcs1_end || lines[i] === pkcs8_end)
      {
        key += lines[i] + '\n';
        inKey = false;
      }
      else if (inKey)
      {
        key += lines[i] + '\n';
      }
    }
    return key;
  }

  function initNoKey()
  {
    $('#main-content').html(noKeyText);
    $('#sa-form').submit(clickGetCert);
    $('#sa-button').click(clickGetCert);
    $('#paste-form').submit(clickPasteCert);
    $('#paste-button').click(clickPasteCert);
    $('#forgetAll').click(clickLogout);
    window.addEventListener('message', messageCert);
    var choice = $('#sa-choice');
    var i = 0;
    var defaultFound = false;

    for (i = 0; i < saList.length; i += 1)
    {
      var attrs = 'value="' + saList[i].url + '"';
      if (defaultMA.url && defaultMA.url == saList[i].url)
      {
        defaultFound = true;
        attrs += ' selected="selected"'
      }
      choice.append('<option ' + attrs + '>' + saList[i].name +
                    '</option>');
    }
    if (! defaultFound && defaultMA.url && defaultMA.name)
    {
      choice.append('<option value="' + defaultMA.url + '" selected="selected">'
                    + defaultMA.name + '</option>');
    }

    $('#default-choice-1').click(function() {
      $('#sa-choice').val('https://www.emulab.net/getsslcertjs.php3');
      $('#sa-button').click();
    });
    $('#default-choice-2').click(function() {
      $('#sa-choice').val('https://portal.geni.net/secure/loadcert.php');
      $('#sa-button').click();
    });
  }

  function initAuthorize()
  {
    var fillId = '';
    if (toolId)
    {
      fillId = toolId;
    }
    var content = authorizeTemplate({id: fillId});
    $('#main-content').html(content);
    if (! toolId)
    {
      $('.tool-info').hide();
    }
    else
    {
      $('.no-tool-info').hide();
    }
    $('#sign').click(clickSign);
    $('#private').submit(clickSign);
    $('#logout').click(clickLogout);
    window.removeEventListener('message', messageCert);
    window.addEventListener('message', messageAck);
    if (! isEncrypted(encryptedKey))
    {
      var shouldAccept = false;
      try {
	var accepted = JSON.parse(window.localStorage.accepted);
	if (accepted[toolId] !== undefined)
	{
	  $('#time').val(accepted[toolId]);
	  shouldAccept = true;
	}
      } catch (e) {}
      if (shouldAccept)
      {
	clickSign();
      }
      $('#password-container').hide();
      $('.password-text').hide();
      $('#rememberCheckContainer').show();
    }
    else
    {
      $('#password-container').show();
      $('.password-text').show();
      $('#rememberCheckContainer').hide();
    }
    getCertFields(certList);

    $('#advancedCheck').change(function() {
      if ($('.toggleAdvanced').hasClass('toggled')) {
        $('.toggleAdvanced').each(function() {
          $(this).removeClass('toggled');
        });
      }
      else {
        $('.toggleAdvanced').each(function() {
          $(this).addClass('toggled');
        });
      }
    });
  }

  /**
   * Populate the KeyInfo block in the signature. Add a KeyValue block
   * with the signing public key info (modulus and exponent). Put the
   * full certificate chain in the X509Data block so that it can be
   * verified.
   *
   * @param chain an one or more certificates in an array where the
   *              first is the signing certificate.
   */
  function GENIKeyInfo(chain)
  {
    this.chain = chain;

    /**
     * Note: This is a private function from forge.rsa
     * Converts a positive BigInteger into 2's-complement big-endian bytes.
     *
     * @param b the big integer to convert.
     *
     * @return the bytes.
     */
    this.bnToBytes = function(b) {
      // prepend 0x00 if first byte >= 0x80
      var hex = b.toString(16);
      if(hex[0] >= '8') {
        hex = '00' + hex;
      }
      return forge.util.hexToBytes(hex);
    }

    this.keyValue = function(pemcert) {
      var cert = parsePemCert(this.chain[0]);
      var modulus = cert.publicKey.n;
      var modulus64 = forge.util.encode64(this.bnToBytes(modulus), 64);
      /* forge puts CR and NL in, we just want NL. */
      modulus64 = modulus64.replace(/\r\n/g,"\n");
      var exponent = cert.publicKey.e;
      var exponent64 = forge.util.encode64(this.bnToBytes(exponent), 64);
      var res = "\n<KeyValue>\n<RSAKeyValue>\n<Modulus>";
      res += modulus64;
      res += "\n</Modulus>\n<Exponent>";
      res += exponent64;
      res += "</Exponent>\n</RSAKeyValue></KeyValue>";
      return res;
    }

    this.getKeyInfo = function(key) {
      var res = "";
      if (this.chain) {
        res += this.keyValue(this.chain[0]);
        res += "\n<X509Data>";
        for (var i = 0; i < this.chain.length; i++) {
          res += "\n<X509Certificate>";
          res += this.chain[i];
          res += "</X509Certificate>";
        }
        res += "\n</X509Data>";
      }
      return res
    };

    this.getKey = function(keyInfo) {
      return "";
    };
  }

    var speaksForCredential;
    var decryptedPrivateKey;
    var userSecret;

  function clickSign(event)
  {
    if (event)
    {
      event.preventDefault();
    }
    if (! isEncrypted(encryptedKey) && $('#rememberCheck')[0].checked)
    {
      try {
	var acceptedStr = window.localStorage.accepted;
	var accepted;
	if (acceptedStr)
	{
	  accepted = JSON.parse(window.localStorage.accepted);
	}
	else
	{
	  accepted = {};
	}
	accepted[toolId] = $('#time').val();
	window.localStorage.accepted = JSON.stringify(accepted);
      } catch (e) { }
    }
    var durationDays = parseInt($('#time').val(), 10);
    var timeOffset = 120 * singleDay;
    if (durationDays && durationDays > 0 && durationDays < 400)
    {
      timeOffset = durationDays * singleDay;
    }
    var e = new Date();
    e.setTime(e.getTime() + timeOffset);
    var eString = e.toISOString();
    var xml = xmlTemplate({ 'expires': eString,
                            'userKeyhash': getKeyhash(certList[0]),
                            'toolKeyhash': getKeyhash(speakerCert) });
    var password = $('#password').val();
    try
    {
      var decrypted = forge.pki.decryptRsaPrivateKey(encryptedKey, password);
	decryptedPrivateKey = decrypted;
      var sig = new SignedXml();
      sig.addReference("//*[local-name(.)='credential']");
      sig.signingKey = decrypted;
      sig.signatureNode = "//*[local-name(.)='signatures']";
      if (certList.length > 0)
      {
        sig.keyInfoProvider = new GENIKeyInfo(certList);
      }
      sig.computeSignature(xml);
	speaksForCredential = sig.getSignedXml();
      var userToken = generateUserToken(speakerCert);
      var data = {
        id: toolId,
        credential: speaksForCredential,
	userToken: userToken,
	userCertificate: wrapCertificates(certList)
      };
      if (window.opener) {
        window.opener.postMessage(data, '*');
	  console.log('posting credential');
      } else {
        // This can't be the only way to redirect with POST, sigh...
        var backForm = document.createElement('form');
        backForm.method = 'POST';
        backForm.action = backTo;
        backForm.name = 'backto';

        var idInput = document.createElement('input');
        idInput.type = 'hidden';
        idInput.name = 'data_id';
        idInput.value = data.id;

        var credentialInput = document.createElement('input');
        credentialInput.type = 'hidden';
        credentialInput.name = 'data_credential';
        credentialInput.value = data.credential;

        backForm.appendChild(idInput);
        backForm.appendChild(credentialInput);
        backForm.submit();
      }
    }
    catch (e)
    {
      console.log('Error signing credential:', e);
      error.show();
      $('#password-container #password').on('input', function() {
        error.hide();
        $('#password-container #password').off('input');
      })
    }
  }

  function clickGetCert(event)
  {
    event.preventDefault();
    if (certWindow)
    {
      certWindow.close();
    }
    var url = $('#sa-choice').val();
    if (url !== '')
    {
      var encodedToolId = encodeURIComponent(toolId);
      url += '?toolid=' + encodedToolId;
      certWindow = window.open(url, 'Slice Authority Credential',
                               'height=400,width=600');
    }
    return false;
  }

  function clickPasteCert(event)
  {
    event.preventDefault();
    var input = $('#paste-input');
    if (input.val() !== '')
    {
      addCert(input.val());
    }
    return false;
  }

  function messageAck(event)
  {
      console.log('messageAck');
      if (event.source === window.opener && event.data && event.data.id &&
	  event.data.id === toolId)
      {
	  if (event.data.ack)
	  {
	      window.close();
	  }
	  else if (event.data.toolToken)
	  {
	      sendAuthorizationToken(event.data.toolToken);
	  }
      }
  }

  function sendAuthorizationToken(toolToken)
  {
      var encryptedCredential = encryptWithSecrets(userSecret, getToolSecret(toolToken));
      var data = {
	  id: toolId,
	  encryptedCredential: encryptedCredential,
	  credential: speaksForCredential
      };
      window.opener.postMessage(data, '*');
      console.log('posting auth token');
  }

  function generateUserToken(speakerCert)
  {
      var secretBytes = forge.random.getBytesSync(32);
      userSecret = forge.util.bytesToHex(secretBytes);
      var p7 = forge.pkcs7.createEnvelopedData();
      var cert = forge.pki.certificateFromPem(wrapCertificates([speakerCert]));
      p7.addRecipient(cert);
      p7.content = forge.util.createBuffer(userSecret);
      p7.encrypt();
      return forge.pkcs7.messageToPem(p7);
  }

  function getToolSecret(toolToken)
  {
      var p7 = forge.pkcs7.messageFromPem(toolToken);
      p7.decrypt(p7.recipients[0], decryptedPrivateKey);
      return p7.content.toString();
  }

  function encryptWithSecrets(userSecret, toolSecret)
  {
      var md = forge.md.sha256.create();
      md.update(speaksForCredential + userSecret + toolSecret);
      return md.digest().toHex();
  }

  function messageCert(event)
  {
      console.log('messageCert');
    if (event.source === certWindow && event.data && event.data.authority)
    {
      if (event.data.certificate)
      {
        addCert(event.data.certificate);
      }
      else if (event.data.sfcred)
      {
        reflectCred(event.data.sfcred);
      }
    }
  }

  function addCert(cert)
  {
    setupClientPem(cert);
    if (isEncrypted(encryptedKey))
    {
      try {
	window.localStorage.certificateList = JSON.stringify([cert]);
      } catch (e) {}
    }
    initAuthorize();
  }

  function setupClientPem(pem)
  {
    encryptedKey = extractKey(pem);
    certList = extractCertificates(pem);
  }

  /*
   * Reflect a speaks-for credential back to the original tool
   */
  function reflectCred(cred)
  {
    var data = {
      id: toolId,
      credential: cred
    };
    window.opener.postMessage(data, '*');
      console.log('posting reflect cred');
    window.removeEventListener('message', messageCert);
    window.addEventListener('message', messageAck);
  }

  function clickLogout(event)
  {
    event.preventDefault();
    try {
      delete window.localStorage.certificateList;
      delete window.localStorage.accepted;
    } catch (e) {}
    certList = [];
    encryptedKey = "";
    initNoKey();
    return false;
  }

  function htmlEncode(value)
  {
    var result = '';
    if (value) {
      result = $('<div />').text(value).html();
    }
    return result;
  };

  function getQueryParams(qs) {
    qs = qs.split("+").join(" ");
    var params = {};
    var re = /[?&]?([^=]+)=([^&]*)/g;
    var tokens = re.exec(qs);

    while (tokens) {
      params[decodeURIComponent(tokens[1])]
        = decodeURIComponent(tokens[2]);
      tokens = re.exec(qs);
    }

    return params;
  }

  function parsePemCert(pem)
  {
    var cert = forge.pki.certificateFromPem('-----BEGIN CERTIFICATE-----\n' +
                                            pem + '\n' +
                                            '-----END CERTIFICATE-----\n');
    return cert;
  }

  function getKeyhash(pem)
  {
    var cert = parsePemCert(pem);
    var asnBitstring = forge.pki.publicKeyToRSAPublicKey(cert.publicKey);
    var derBitstring = forge.asn1.toDer(asnBitstring);

    var sha1 = forge.md.sha1.create();
    sha1.update(derBitstring.bytes());
    return sha1.digest().toHex();
  }

  function isEncrypted(key)
  {
    var result = true;
    try
    {
      var decrypted = forge.pki.decryptRsaPrivateKey(key, '');
      // If there are no exceptions, this is an unencrypted key.
      result = false;
    }
    catch (e) {}
    return result;
  }

  function getCertFields (certificateList)
  {
//    _.each(certificateList, function (certificate) {
      var cert = parsePemCert(certificateList[0]);
      var found = false;
      var urn;
      var name;

      var altField;
      _.each(cert.extensions, function (field) {
	if (field.name === 'subjectAltName')
	{
	  altField = field;
	}
      });
      if (altField)
      {
	_.each(altField.altNames, function (item) {
	  if (item.type === 6 && item.value.substr(0, 12) === 'urn:publicid')
	  {
	    urn = item.value;
	  }
	});
      }
      if (urn)
      {
	name = urn.split('+')[3];
	found = true;
      }

      if (found)
      {
	var authorityUrn = 'urn:publicid:IDN+' + urn.split('+')[1] +
	  '+authority+sa';

	fillUserForm(name, authorityUrn);
      }
//    });
  }

  function fillUserForm(name, authority)
  {
    $('#userName').val(name);
    $('#authorityName').val(authority.split('+')[1]);
  }

  $(document).ready(initialize);
});
